File: members/U02BQ6G1SQJ/programmers/problem/크레인 인형뽑기 게임.py
REMOVE_CNT = 2;

def moveDoll(transposedBoard : list, stack : list, targetRow : int) -> None:
    targetCol = -1;
    
    for curCol in range(len(transposedBoard)):
        if (transposedBoard[targetRow][curCol] != 0):
            targetCol = curCol;
            break;
    
    if (targetCol == -1):
        return;
    
    stack.append(transposedBoard[targetRow][targetCol]);
    transposedBoard[targetRow][targetCol] = 0;
    
    # print(transposedBoard);
    # print(stack);
    # print(targetRow, targetCol);
    
    return None;

def checkStack(stack : list) -> int:
    if ((len(stack) < REMOVE_CNT) or (stack[-1] != stack[-2])):
        return 0;
    
    stack.pop();
    stack.pop();
    return REMOVE_CNT;

def solution(board : list, moves : list) -> int:
    answer = 0;
    (stack, transposedBoard) = ([], list(map(list, zip(*board))));
    
    # print(transposedBoard);
    
    for curMove in moves:
        moveDoll(transposedBoard, stack, curMove - 1);
        answer += checkStack(stack);
    
    return answer;

File: members/U02BWCHJA6Q/Programers/kakao_internship_2019/크레인_인형뽑기_게임.js
function solution(board, moves) {
  let deletedCount = 0;
  const map = new Map();
  const stack = [];

  for (let i = 0; i < board.length; i++) {
    map.set(i + 1, []);
  }

  for (let y = 0; y < board.length; y++) {
    for (let x = 0; x < board.length; x++) {
      if (board[y][x] !== 0) {
        map.get(x + 1).push(board[y][x]);
      }
    }
  }

  moves.forEach(index => {
    const doll = map.get(index).shift();

    if (doll) {
      if (stack.length === 0) {
        stack.push(doll);

        return;
      }

      if (stack[stack.length - 1] === doll) {
        stack.pop();
        deletedCount++;
      } else {
        stack.push(doll);
      }
    }
  });

  return deletedCount * 2;
}
